### PR TITLE
Fix: Use fallback if RoutePattern is MVC.

### LINKED ISSUE #1189: Fix performance transaction name when using MVC routes
When using `MapControllerRoute`, the SDK is currently just using the route template regardless of the endpoint. I.e:

image

We need to change the logic:

https://github.com/getsentry/sentry-dotnet/blob/a12bb9f2e0748c4560b207815631bfd5be651ac5/src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs#L19-L39

To account for such template.
For example:

```csharp
if (pattern.Contains("{controller=")) { 
  Regex.Replace("\{controller=(.*)\}", controllerName)
}
if (pattern.Contains("{action=")) { 
  Regex.Replace("\{action=(.*)\}", controllerName)
}
if (pattern.Contains("{area=")) { 
  Regex.Replace("\{area=(.*)\}", controllerName)
}
```

Such as:
> area/{controller=MyController}/bla/{action=MyAction}/{id?}/blabla

Becomes: 
> area/Controller/bla/Action/{id?}/blabla

### PR DESCRIPTION
**Pull request by lucas-zimerman**
Despite a context feature having a RouteEndpoint, it doesn't mean we can use the RawText value of the RoutePattern since it may result in a default route for different paths.
For a simple fix, I opted to just use the fallback to get the MVC route since we didn't have any proper data of the Route Template on the given Endpoint.

Additionally, that the SDK is creating transactions for items like loading a CSS file, thus results in Unknown Routes. If desired, we could use context.Request.Path.Value as a last fallback for cases like that.

Sample:
image

Also, the issue can be easily reproduced on the sample Sentry.Samples.AspNetCore.Mvc

Close #1189.

### CODE REVIEW

File: test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+        [Theory]
+        [InlineData("{area=MyArea}/{controller=Home}/{action=Index}/{id?}", "theArea/house/about/{id?}", "house", "about", "theArea")]
+        [InlineData("{area=MyArea}/{controller=Home}/{action=Index}/{id?}", "{area=MyArea}/house/about/{id?}", "house", "about", null)]
+        [InlineData("{controller=Home}/{action=Index}/{id?}", "house/about/{id?}", "house", "about", null)]
+        [InlineData("{controller=Home}/{action=Index}", "house/about", "house", "about", null)]
+        [InlineData("{controller=Home}/{id?}", "house/{id?}", "house", "about", null)]
+        [InlineData("{action=Index}/{id?}", "about/{id?}", null, "about", null)]
+        [InlineData("not/mvc/", "not/mvc/", "house", "about", "area")]
+        public void ReplaceMcvParameters_ParsedParameters(string routeInput, string assertOutput, string context, string action, string area)
```
Comment (bruno-garcia) [inline test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs:27]: Nice approach for testing. But missing some fixed bits in the URL like `abc/{controller=}`

And perhaps also: `abc/{controller}` <- this isn't an MVC format, it's just a parameter called controller.

File: test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Sentry.AspNetCore.Extensions;
+using Xunit;
+
+namespace Sentry.AspNetCore.Tests.Extensions
+{
+    public class HttpContextExtensionsTests
+    {
+#if !NETCOREAPP2_1
```
Comment (bruno-garcia) [inline test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs:10]: Can this be defined the whole file instead?

File: test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        public void ReplaceMcvParameters_ParsedParameters(string routeInput, string assertOutput, string context, string action, string area)
+        {
+            // Arrange
+            var httpContext = new DefaultHttpContext();
+            AddRouteValuesIfNotNull(httpContext.Request.RouteValues, "controller", context);
+            AddRouteValuesIfNotNull(httpContext.Request.RouteValues, "action", action);
+            AddRouteValuesIfNotNull(httpContext.Request.RouteValues, "area", area);
+
+            // Act
+            var filteredRoute = HttpContextExtensions.ReplaceMcvParameters(routeInput, httpContext);
```
Comment (bruno-garcia) [inline test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs:38]: Mcv -> Mvc

File: src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs [hunk 4]
```diff
@@ -43,6 +47,34 @@ internal static class HttpContextExtensions
...
+            {
+                route = Regex.Replace(route, "\\{area=[^\\}]+\\}", area);
+            }
+
+            return route;
+        }
+
+        // Internal for testing.
+        internal static bool RouteHasMvcParameters(string route)
+            => route.Contains("{controller=") || route.Contains("{action=") || route.Contains("{area=");
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs:77]:
```suggestion
        internal static bool RouteHasMvcParameters(string route)
            => route.Contains("{controller=") || route.Contains("{action=") || route.Contains("{area=");

```

File: src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs [hunk 5]
```diff
@@ -43,6 +47,34 @@ internal static class HttpContextExtensions
...
+            }
+
+            if (routeData?.Values["action"]?.ToString() is { } action)
+            {
+                route = Regex.Replace(route, "\\{action=[^\\}]+\\}", action);
+            }
+
+            if (routeData?.Values["area"]?.ToString() is { } area)
+            {
+                route = Regex.Replace(route, "\\{area=[^\\}]+\\}", area);
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs:69]: Please remove the `\\`:

```suggestion
                route = Regex.Replace(route, "{area=[^}]+}", area);
```

File: test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs (hunk 2 above)
Comment (lucas-zimerman) [reply test/Sentry.AspNetCore.Tests/Extensions/HttpContextExtensionsTests.cs:10]: sure

File: src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs [hunk 6]
```diff
@@ -11,14 +12,17 @@ internal static class HttpContextExtensions
...
             // Requires .UseRouting()/.UseEndpoints()
             var endpoint = context.Features.Get<IEndpointFeature?>()?.Endpoint as RouteEndpoint;
             var routePattern = endpoint?.RoutePattern.RawText;
 
             if (!string.IsNullOrWhiteSpace(routePattern))
             {
-                return routePattern;
+                // Skip route pattern if it resembles to a MVC route or null  e.g.
+                // {controller=Home}/{action=Index}/{id?}
+                return RouteHasMvcParameters(routePattern) ? ReplaceMvcParameters(routePattern, context) : routePattern;
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/Extensions/HttpContextExtensions.cs:25]:
```suggestion
                return RouteHasMvcParameters(routePattern) 
                    ? ReplaceMvcParameters(routePattern, context)
                    : routePattern;
```

**Review by bruno-garcia — APPROVED** [review]

**Pull request merged**